// betterryver/index.js
/* BetterRyver (working name)
 * A Ryver enhancement suite, loosely heavily inspired by BetterDiscord
 */

"use strict";

const _path = require("path");
const _fs   = require("fs");

const BetterRyverUI = require("./ui");

module.exports = class BetterRyverLoader {
    constructor(window) {
        this.window = window;
        this.internalConfig = require("./config.json");
        this.safeMode = this.internalConfig.loader.forceSafeMode;
        this.config = {};
        this.ui = new BetterRyverUI(window);
    }

    init() {
        this.determineDataDir();
        this.loadConfig();
    }

    determineDataDir() {
        if (process.platform == "win32") {
            this.internalConfig.dataDir = _path.join(process.env.APPDATA, "BetterRyver/");
        } else if (process.platform == "darwin") {
            this.internalConfig.dataDir = _path.join(process.env.HOME, "/Library/Preferences/BetterRyver/");
        } else {
            this.internalConfig.dataDir = _path.join(process.env.HOME, ".betterryver/");
        }

        if (!_fs.existsSync(this.internalConfig.dataDir)) {
            _fs.mkdirSync(this.internalConfig.dataDir);
        }
    }

    loadConfig() {
        try {
            this.config = require(_path.join(this.internalConfig.dataDir, "config.json"));
        } catch (e) {
            this.internalConfig.safeMode = true;
            let errMsg = this.ui.errorPopup("BetterRyver could not load its config. Safe mode enabled.")
            this.ui.showPopup(errMsg);
        }
    }

    saveConfig() {
        let jsonString = JSON.stringify(this.config);
        let writeStream = _fs.createWriteStream(_path.join(this.internalConfig.dataDir, "config.json"));
        writeStream.write(jsonString, "utf8", () => {
            writeStream.end();
        });
    }
}

class BetterRyverAPI {
    constructor(config) {
        this.internalConfig = config
    }
}
