// betterryver/index.js
/* BetterRyver (working name)
 * A Ryver enhancement suite, loosely heavily inspired by BetterDiscord
 */

"use strict";

const _path = require("path");
const _fs = require("fs");

const BetterRyverUI = require("./ui");
const BetterRyverKeybinds = require("./keybinds");

module.exports = class BetterRyverLoader {
    constructor(window) {
        window.betterryver = this;

        this.window = window;
        this.webviews = document.getElementsByTagName("webview");

        this.internalConfig = require("./config.json");
        this.safeMode = this.internalConfig.loader.forceSafeMode;
        this.config = {};

        this.ui = new BetterRyverUI(window);
        this.keybinds = new BetterRyverKeybinds(window);
    }

    init() {
        this.determineDataDir();
        this.loadConfig();
        this.injectCSS();
    }

    determineDataDir() {
        if (process.platform == "win32") {
            this.internalConfig.dataDir = _path.join(process.env.APPDATA, "BetterRyver/");
        } else if (process.platform == "darwin") {
            this.internalConfig.dataDir = _path.join(process.env.HOME, "/Library/Preferences/BetterRyver/");
        } else {
            this.internalConfig.dataDir = _path.join(process.env.HOME, ".betterryver/");
        }

        if (!_fs.existsSync(this.internalConfig.dataDir)) {
            _fs.mkdirSync(this.internalConfig.dataDir);
        }
    }

    loadConfig() {
        try {
            this.config = require(_path.join(this.internalConfig.dataDir, "config.json"));
            this.ui.showPopup(this.ui.infoPopup("BetterRyver configuration loaded."));
        } catch (e) {
            this.internalConfig.safeMode = true;
            this.ui.showPopup(this.ui.errorPopup("BetterRyver could not load its config. Safe mode enabled.", true));
        }
    }

    saveConfig() {
        let jsonString = JSON.stringify(this.config);
        let writeStream = _fs.createWriteStream(_path.join(this.internalConfig.dataDir, "config.json"));
        writeStream.write(jsonString, "utf8", () => {
            writeStream.end();
        });
    }

    injectCSS() {
        for (let orgName in this.config.orgs) {
            let orgWebview = document.getElementById("webview_" + orgName);
            let enabledCss = this.config.orgs[orgName].enabledCss;
            if (orgWebview == null) {
                if (!(orgName == "_template")) console.log("Could not find webview_" + orgName);
                continue;
            }
            
            orgWebview.addEventListener("did-navigate", () => {
                let _this = this;
                for (let css in enabledCss) {
                    _fs.readFile(_path.join(this.internalConfig.dataDir, "css", enabledCss[css]), "utf-8", (err, data) => {
                        if (err) _this.ui.showPopup(_this.ui.warnPopup(err));
                        orgWebview.insertCSS(data);
                    });
                }
            });
        }
    }

    getCurrentOrganisation() {
        let highestIndex = 0;
        let highest;
        let webviews = document.getElementsByTagName("webview");
        for (let wv of webviews) {
            let zindex = window.getComputedStyle(wv.parentNode).getPropertyValue("z-index");
            if (!isNaN(zindex)) if (parseInt(zindex) > highestIndex) {
                highestIndex = parseInt(zindex);
                highest = wv;
            }
        }
        return highest;
    }
}

class BetterRyverAPI {
    constructor(config) {
        this.internalConfig = config
    }
}
