// betterryver/ui.js
const fs = require("fs");

module.exports = class BetterRyverUI {
    constructor(window) {
        this.window = window;

        this.injectCss(fs.readFileSync("./resources/app/node_modules/betterryver/ui.css"));
    }

    injectCss(cssString) {
        let css = document.createElement("style");
        css.innerHTML = cssString;
        document.getElementsByTagName("head")[0].appendChild(css);
    }

    popup(message, dismissible) {
        let div = document.createElement("div");
        // div.style.display = "block";
        // div.style.position = "absolute";
        // div.style.top = "0px";
        // div.style.padding = "10px";
        // div.style.fontFamily = "'Open Sans', sans-serif";
        // div.style.zIndex = "200";
        // div.style.width = "100%";
        // div.style.transform = "translate3d(-100%, 0px, 0px)";
        div.classList.add("popup");
        if (dismissible) {
            let _this = this;
            message = message + " <a href='#' class='popup-dismiss'>Dismiss</a>";
            div.addEventListener("click", () => { _this.removePopup(div); });
        }
        div.innerHTML = message;
        document.getElementsByTagName("body")[0].insertBefore(div, document.getElementById("appMountPoint"));
        return div;
    }

    errorPopup(message, dismissible) {
        let div = this.popup(message, dismissible);
        // div.style.backgroundColor = "#ffc3c3";
        // div.style.borderColor = "#dd4646";
        // div.style.borderBottomWidth = "2px";
        div.classList.add("popup-error");
        // document.getElementsByTagName("body")[0].insertBefore(div, document.getElementById("appMountPoint"));
        return div;
    }

    showPopup(div) {
        div.style.transform = "translateY(0%)";
    }

    hidePopup(div) {
        div.style.transform = "translateY(-100%)";
    }

    removePopup(div) {
        this.hidePopup(div);
        setTimeout(() => {div.remove()}, 500)
    }
}
